gen/constant.go

```diff
@@ -20,17 +20,222 @@
 
 package gen
 
-import "github.com/thriftrw/thriftrw-go/compile"
+import (
+	"fmt"
+	"strconv"
 
-// Constant TODO
+	"github.com/thriftrw/thriftrw-go/compile"
+)
+
+// Constant generates code for `const` expressions in Thrift files.
 func Constant(g Generator, c *compile.Constant) error {
 	err := g.DeclareFromTemplate(
-		`
-		const <goCase .Name> <typeReference .Type> = nil // TODO
-		`,
+		`<if canBeConstant .Type>const<else>var<end> <goCase .Name> <typeReference .Type> = <constantValue .Value .Type>`,
 		c,
+		TemplateFunc("constantValue", ConstantValue),
+		TemplateFunc("canBeConstant", canBeConstant),
 	)
-	// TODO(abg): Implement constantValue
-	// TODO(abg): Complex types will need to use var rather than const
 	return wrapGenerateError(c.Name, err)
 }
+
+// ConstantValue generates an expression containing the given constant value of
+// the given type.
+//
+// The constant must already have been linked to the given type.
+func ConstantValue(g Generator, c compile.ConstantValue, t compile.TypeSpec) (string, error) {
+	switch v := c.(type) {
+	case compile.ConstantBool:
+		if v {
+			return "true", nil
+		}
+		return "false", nil
+	case compile.ConstantDouble:
+		return fmt.Sprint(float64(v)), nil
+	case compile.ConstantInt:
+		return fmt.Sprint(int(v)), nil
+	case compile.ConstantList:
+		l, ok := t.(*compile.ListSpec)
```

Comment:
can we break the more complex constant generation things into their own functions. there doesn't seem to be any advantage to having them inline here. maybe the type check can be done in this function, but then you can call a function to do the rest?